File: CS/TestMyButtonEdit/MyButtonEditPainter.cs
using System;
using System.Drawing;
using DevExpress.Utils;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using DevExpress.Skins;




namespace DevExpress.MyControl
{
    public class MyButtonEditPainter : ButtonEditPainter
    {
        private int CalcButtonsWidth(ControlGraphicsInfoArgs info)
        {
            ButtonEditViewInfo vi = info.ViewInfo as ButtonEditViewInfo;
            int iResult = 0;

            if (vi.RightButtons.Count > 0)
            {
                for (int i = 0; i < vi.RightButtons.Count; i++)
                    iResult += vi.RightButtons[i].Bounds.Width;

            }
            return iResult;
        }

        protected override void DrawString(ControlGraphicsInfoArgs info, Rectangle bounds, string text, AppearanceObject appearance)
        {
            ButtonEditViewInfo vi = info.ViewInfo as ButtonEditViewInfo;


            int buttonsWidth = CalcButtonsWidth(info);

            int text_width = vi.Bounds.Width - buttonsWidth;
            if (text_width <= 1)
                text_width = vi.ClientRect.Width;


            bounds = new Rectangle(bounds.X, bounds.Y, text_width, bounds.Height);
            appearance.DrawString(info.Cache, text, bounds, appearance.GetForeBrush(info.Cache), appearance.GetTextOptions().GetStringFormat(info.ViewInfo.DefaultTextOptions));
        }



        protected override void DrawButton(ButtonEditViewInfo viewInfo, EditorButtonObjectInfoArgs info)
        {
            if (info.Bounds.Width == 0 || viewInfo.RightButtons.Count == 0) return;
            if (info.Button.Index == 1)
            {
                base.DrawButton(viewInfo, info);
                string str_value = viewInfo.EditValue as string;
                Color captionColor = EditorsSkins.GetSkin(viewInfo.LookAndFeel.ActiveLookAndFeel)[EditorsSkins.SkinEditorButton].Color.GetForeColor();
                info.Cache.DrawString(str_value, viewInfo.Appearance.Font, new SolidBrush(captionColor), info.Bounds, info.Appearance.GetStringFormat());
                return;
            }

            base.DrawButton(viewInfo, info);
        }
    }
}

File: VB/TestMyButtonEdit/MyButtonEditPainter.vb
Imports System
Imports System.Drawing
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.Skins




Namespace DevExpress.MyControl
    Public Class MyButtonEditPainter
        Inherits ButtonEditPainter

        Private Function CalcButtonsWidth(ByVal info As ControlGraphicsInfoArgs) As Integer
            Dim vi As ButtonEditViewInfo = TryCast(info.ViewInfo, ButtonEditViewInfo)
            Dim iResult As Integer = 0

            If vi.RightButtons.Count > 0 Then
                For i As Integer = 0 To vi.RightButtons.Count - 1
                    iResult += vi.RightButtons(i).Bounds.Width
                Next i

            End If
            Return iResult
        End Function

        Protected Overrides Sub DrawString(ByVal info As ControlGraphicsInfoArgs, ByVal bounds As Rectangle, ByVal text As String, ByVal appearance As AppearanceObject)
            Dim vi As ButtonEditViewInfo = TryCast(info.ViewInfo, ButtonEditViewInfo)


            Dim buttonsWidth As Integer = CalcButtonsWidth(info)

            Dim text_width As Integer = vi.Bounds.Width - buttonsWidth
            If text_width <= 1 Then
                text_width = vi.ClientRect.Width
            End If


            bounds = New Rectangle(bounds.X, bounds.Y, text_width, bounds.Height)
            appearance.DrawString(info.Cache, text, bounds, appearance.GetForeBrush(info.Cache), appearance.GetTextOptions().GetStringFormat(info.ViewInfo.DefaultTextOptions))
        End Sub



        Protected Overrides Sub DrawButton(ByVal viewInfo As ButtonEditViewInfo, ByVal info As EditorButtonObjectInfoArgs)
            If info.Bounds.Width = 0 OrElse viewInfo.RightButtons.Count = 0 Then
                Return
            End If
            If info.Button.Index = 1 Then
                MyBase.DrawButton(viewInfo, info)
                Dim str_value As String = TryCast(viewInfo.EditValue, String)
                Dim captionColor As Color = EditorsSkins.GetSkin(viewInfo.LookAndFeel.ActiveLookAndFeel)(EditorsSkins.SkinEditorButton).Color.GetForeColor()
                info.Cache.DrawString(str_value, viewInfo.Appearance.Font, New SolidBrush(captionColor), info.Bounds, info.Appearance.GetStringFormat())
                Return
            End If

            MyBase.DrawButton(viewInfo, info)
        End Sub
    End Class
End Namespace
